ttnn/cpp/ttnn/operations/embedding/device/kernels/dataflow/embeddings.cpp

```diff
@@ -3,145 +3,68 @@
+    uint32_t offset = weights_offset;
+    uint32_t index = index_idx;
 
-        noc_async_read_barrier();
-    }
-#endif
+    uint64_t noc_input_src_addr = get_noc_addr(curr_row, input) + offset;
+    noc_async_read<input_block_size_bytes>(noc_input_src_addr, input_l1_addr, input_block_size_bytes);
+    noc_async_read_barrier();
 
-    cb_reserve_back(cb_id_in1, 1);
-    uint32_t input_l1_addr = get_write_ptr(cb_id_in1);
-#if defined BFP16
-    volatile tt_l1_ptr uint16_t* input_l1_ptr = reinterpret_cast<volatile tt_l1_ptr uint16_t*>(input_l1_addr);
-#else
-    volatile tt_l1_ptr uint32_t* input_l1_ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(input_l1_addr);
-#endif
-    auto read_block = [&](const uint32_t& token_idx, const uint32_t& width_size) {
+    for (uint32_t i = 0; i < num_blocks; ++i) {
         cb_reserve_back(cb_id_in0, 1);
         uint32_t l1_write_addr = get_write_ptr(cb_id_in0);
-        uint64_t src_noc_addr;
-        uint32_t token = input_l1_ptr[token_idx];
-#if defined PADDED
-        if (token == pad_token) {
-            src_noc_addr = pad_noc_addr;
-        } else {
-            src_noc_addr = get_noc_addr(token, weights);
-        }
-#elif defined BINARY
-        if (token == 0) {
-            src_noc_addr = zero_noc_addr;
-        } else {
-            src_noc_addr = one_noc_addr;
-        }
-#else
-#if defined BFP16
-        union {
-            float f;
-            uint32_t u;
-        } u;
-        u.u = (uint32_t)input_l1_ptr[token_idx] << 16;
-        uint32_t token_casted = static_cast<uint32_t>(u.f);
-        src_noc_addr = get_noc_addr(token_casted, weights);
-#else
-        src_noc_addr = get_noc_addr(token, weights);
-#endif
-#endif
-        noc_async_read(src_noc_addr, l1_write_addr, width_size);
+        input_token_t token = input_l1_ptr[index];
+        uint64_t src_noc_addr = get_token_noc_addr(token, weights);
+        noc_async_read<weight_stick_size>(src_noc_addr, l1_write_addr, weight_stick_size);
         noc_async_read_barrier();
         cb_push_back(cb_id_in0, 1);
-    };
 
-    uint32_t curr_row = batch_offset;
-    uint32_t offset = weights_offset;
-    uint32_t index = index_idx;
-    bool read_indices = true;
-    for (uint32_t i = 0; i < num_blocks; ++i) {
-        if (read_indices) {
-            uint64_t noc_input_src_addr = get_noc_addr(curr_row, input) + offset;
-            noc_async_read(noc_input_src_addr, input_l1_addr, input_block_size_bytes);
-            noc_async_read_barrier();
-            read_indices = false;
-        }
-        read_block(index, weight_stick_size);
         index++;
         if (index == rows_per_block) {
             index = 0;
-            read_indices = true;
             offset += input_block_size_bytes;
             if (offset == input_page_size) {
                 offset = 0;
                 curr_row++;
             }
+            if (i != num_blocks - 1) {
+                noc_input_src_addr = get_noc_addr(curr_row, input) + offset;
```

Comment:
Are there any alignment concerns?

Reply:
I don't believe so. I haven't changed the functionality of this kernel, this just moved things around/removed a boolean. On host the block/offset is calculated to be aligned.